**Objective**
- Find and address mixed type variables in a dataframe
- Find and address missing values
- Find and address duplicate values


In [2]:
import pandas as pd
import numpy as np 
import os

In [3]:
# create a string folderpath
folderpath = r'C:\Users\Bamidele\Desktop\Training_content\March 2025 - Data Analytics Cohort\Python\Istacart Basket Project'

In [4]:
# import the dataset as a dataframe using the folder os.path.join
df_ords = pd.read_csv(os.path.join(folderpath, '02 Data', 'Original Data', 'orders.csv'), index_col= False )
df_prods = pd.read_csv(os.path.join(folderpath, '02 Data', 'Original Data', 'products.csv'), index_col= False )

In [5]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

## Mixed -Type Data

In [7]:
# create a dataframe
df_test2 = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test2['mix'] = ['a', 'b', '1', 'true' ]

In [9]:
df_test2.head()

mix
0     a
1     b
2     1
3  true

In [10]:
# create a dataframe
df_test = pd.DataFrame()

In [11]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True ]

In [12]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [13]:
# check for mixed types 

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\Bamidele\AppData\Local\Temp\ipykernel_13088\3989843499.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [14]:
# change the data type of the column
df_test['mix'] = df_test['mix'].astype('str')

In [15]:
# check for mixed types 

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

C:\Users\Bamidele\AppData\Local\Temp\ipykernel_13088\3989843499.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [16]:
# df_test.columns.tolist()

In [31]:
# check for mixed types 

for col in df_test2.columns.tolist():
  weird = (df_test2[[col]].applymap(type) != df_test2[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test2[weird]) > 0:
    print (col)

C:\Users\Bamidele\AppData\Local\Temp\ipykernel_13088\289779594.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test2[[col]].applymap(type) != df_test2[[col]].iloc[0].apply(type)).any(axis = 1)


## Missing Values

In [29]:
# income 
# debt

# debt/income = derived

### Finding missing values

In [33]:
# check for missing values df_prods dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [35]:
# extracting the missing valueson the product_name
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [37]:
# print df_nan
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### Addressing missing values

In [ ]:
'''
1. Create new variable (Flag)
2. Impute a value (mean or median if the variable is numeric)
3. Remove or filter them out
                   '''

In [ ]:
# 2, 3, 4, 5,5,5,5,5 6,6 and 77

In [ ]:
# df['columnamewithmissing'].fillna(mean value, inplace=True  )
# df['columnamewithmissing'].fillna(median value, inplace=True  )

In [39]:
df_prods.shape

(49693, 5)

In [41]:
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [43]:
df_prods_clean.shape

(49677, 5)

The 16 null entries in the product_name column has been removed, we now have a total of **49677** from **49693**

In [ ]:
# drop na from entire column
''' df_prods.dropna(inplace= True)'''

In [ ]:
# Drop na from a specific column
''' df_prods.dropna(subset= ['product_name'], inplace=True)''''

### Duplicates

In [45]:
df_prods_clean.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [47]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [49]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [51]:
df_prods_clean.shape

(49677, 5)

In [53]:
# remove all duplicates in the dataframe
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [55]:
# Confirm the shape of the dataframe
df_prods_clean_no_dups.shape

(49672, 5)

We now have **49672** rows in our dataframe. The **5** duplicates have been removed

## Export and Save dataframe

In [57]:
# Export your cleaned dataframe to the prepared folder
df_prods_clean_no_dups.to_csv(os.path.join(folderpath, '02 Data', 'Prepared Data', 'products_checked.csv')  )

In [25]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### Tasks

#### 1. Run DESCRIBE FUNCTION ON ORDERS DATAFRAME

In [59]:
#### 1. rUN dESCRIBE FUNCTION ON ORDERS DATAFRAME
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

**'days_since prior order'**: Has a count less than all other rows inndicating missing values. It also has a measure of zero indicating that some orders could be places within the same day. The max value is 30, which may indicate that orders are capped at 30 days or 1 month.

**'order_number'**: Has a max value of 100, which may mean orders are capped at 100 items.

**'order_day of the week'**: Has min value of 0 and max value of 6 meaning the seven numbers (0-6) represent days of the week.

**'order_hour_of_day'**: Has a min value of 0 and max value of 23 meaning the twenty-four numbers (0-23) represent 24 hrs of the day.

#### 2. check for mixed data type

In [67]:
# check for mix data types in the orders dataframe
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\Bamidele\AppData\Local\Temp\ipykernel_13088\2194524189.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Bamidele\AppData\Local\Temp\ipykernel_13088\2194524189.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Bamidele\AppData\Local\Temp\ipykernel_13088\2194524189.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Bamidele\AppData\Local\Temp\ipykernel_13088\2194524189.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\Bamidel

- Since nothing was printed, hence there is no mixed data type in the dataframe

#### 3. check for missing values

In [74]:
# check for missing values in the orders dataframe
df_ords.isnull().sum()

order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

- Unfortunatelyy, there are no missing values in the orders dataframe except for days_since_prior_order with missing value of **206209**

#### 4. Addressing missing values
- Create a subset to view the missing values
- Replace it with median value
- check the dataframe to see that the missing values have been replaced

In [83]:
df_ords[df_ords['days_since_prior_order'].isnull()==True]

order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
39        3343014        4    prior             1          6   
45        2717275        5    prior             1          3   
...           ...      ...      ...           ...        ...   
3420930    969311   206205    prior             1          4   
3420934   3189322   206206    prior             1          3   
3421002   2166133   206207    prior             1          6   
3421019   2227043   206208    prior             1          1   
3421069   3154581   206209    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [85]:
# fill the missing values
df_ords['days_since_prior_order'].fillna(7, inplace= True )

C:\Users\Bamidele\AppData\Local\Temp\ipykernel_13088\1862067724.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_ords['days_since_prior_order'].fillna(7, inplace= True )


In [87]:
# conform if there still missing values
df_ords[df_ords['days_since_prior_order'].isnull()==True]

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [104]:
# check for missing values in the orders dataframe
df_ords.isnull().sum()

order_id                  0
user_id                   0
eval_set                  0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [91]:
# confirm only the filled value
df_ords[df_ords['days_since_prior_order']==7]


order_id  user_id eval_set  order_number  order_dow  \
0         2539329        1    prior             1          2   
11        2168274        2    prior             1          2   
26        1374495        3    prior             1          1   
31        1972919        3    prior             6          0   
32        1839752        3    prior             7          0   
...           ...      ...      ...           ...        ...   
3421052   2017995   206208    prior            34          6   
3421067   1882108   206208    prior            49          1   
3421069   3154581   206209    prior             1          3   
3421070   1889163   206209    prior             2          3   
3421081   2977660   206209    prior            13          1   

         order_hour_of_day  days_since_prior_order  
0                        8                     7.0  
11                      11                     7.0  
26                      14                     7.0  
31                      16                     7.0  
32                      15                     7.0  
...                    ...                     ...  
3421052                 15                     7.0  
3421067                 22                     7.0  
3421069                 11                     7.0  
3421070                 17                     7.0  
3421081                 12                     7.0  

[526817 rows x 7 columns]

Great. Missing values have all been replaced with the median value

#### 5. Duplicates value check

In [97]:
# confirm duplicates
df_ords[df_ords.duplicated()]

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

In [99]:
# Check for the total duplicates in the datasets
df_ords.duplicated().sum()

0

This shows that there are no duplicates in the orders dataframe

#### 6. Export the data

In [112]:
# Export your cleaned dataframe to the prepared folder
df_ords.to_csv(os.path.join(folderpath, '02 Data', 'Prepared Data', 'orders_checked.csv')  )

In [114]:
df_ords.shape

(3421083, 7)